libp2p/transports/wstransport.nim

```diff
@@ -129,6 +137,123 @@ type WsTransport* = ref object of Transport
 proc secure*(self: WsTransport): bool =
   not (isNil(self.tlsPrivateKey) or isNil(self.tlsCertificate))
 
+proc connHandler(
+    self: WsTransport, stream: WSSession, secure: bool, dir: Direction
+): Future[Connection] {.async: (raises: [CatchableError]).} =
+  ## Returning CatchableError is fine because we later handle different exceptions.
+
+  let (observedAddr, localAddr) =
+    try:
+      let
+        codec =
+          if secure:
+            MultiAddress.init("/wss")
+          else:
+            MultiAddress.init("/ws")
+        remoteAddr = stream.stream.reader.tsource.remoteAddress
+        localAddr = stream.stream.reader.tsource.localAddress
+
+      (
+        MultiAddress.init(remoteAddr).tryGet() & codec.tryGet(),
+        MultiAddress.init(localAddr).tryGet() & codec.tryGet(),
+      )
+    except CatchableError as exc:
+      trace "Failed to create observedAddr or listenAddr", description = exc.msg
+      if not (isNil(stream) and stream.stream.reader.closed):
+        safeClose(stream)
+      raise exc
+
+  let conn = WsStream.new(stream, dir, Opt.some(observedAddr), Opt.some(localAddr))
+
+  self.connections[dir].add(conn)
+  proc onClose() {.async: (raises: []).} =
+    await noCancel conn.session.stream.reader.join()
+    self.connections[dir].keepItIf(it != conn)
+    trace "Cleaned up client"
+
+  asyncSpawn onClose()
+  return conn
+
+proc handshakeWorker(
+    self: WsTransport, finished: Future[HttpRequest], secure: bool
+) {.async: (raises: []).} =
+  try:
+    let req = await finished
+    try:
+      let wstransp = await self.wsserver.handleRequest(req).wait(self.handshakeTimeout)
+      let conn = await self.connHandler(wstransp, secure, Direction.In)
+      try:
+        self.acceptResults.addLastNoWait(AcceptResult.ok(conn))
+      except AsyncQueueFullError:
+        await noCancel req.stream.closeWait()
+    except CatchableError as exc:
+      await noCancel req.stream.closeWait()
+      try:
+        self.acceptResults.addLastNoWait(AcceptResult.err(exc))
+      except AsyncQueueFullError:
+        discard
+  except CatchableError as exc:
+    try:
+      self.acceptResults.addLastNoWait(AcceptResult.err(exc))
+    except AsyncQueueFullError:
+      discard
```

Comment:
please improve this logic:
- use specific errors instead of `CatchableError`
- do not use nasted error handling (`try-except` within `try-except`)

for example:

```
let wstransp = 
  try  
    await self.wsserver.handleRequest(req).wait(self.handshakeTimeout)
  except ....

 let conn = 
  try
     await self.connHandler(wstransp, secure, Direction.In)
  except ....
  
```